swaybar/tray/dbusmenu.c

```diff
@@ -0,0 +1,1381 @@
+	return CAIRO_STATUS_SUCCESS;
+}
+
+static cairo_surface_t *read_png(const void *data, size_t data_size) {
+	struct png_stream *png_stream = malloc(sizeof(struct png_stream));
+	if (png_stream == NULL) {
+		sway_log(SWAY_ERROR, "Allocation for PNG stream failed");
+		return NULL;
+	}
+	png_stream->data = data;
+	png_stream->left = data_size;
+	cairo_surface_t *surface =
+		cairo_image_surface_create_from_png_stream(read_png_stream, png_stream);
+	free(png_stream);
+
+	if (cairo_surface_status(surface) == CAIRO_STATUS_SUCCESS) {
+		return surface;
+	}
+
+	cairo_surface_destroy(surface);
+	return NULL;
+}
+
+static int about_to_show_callback(sd_bus_message *msg, void *data,
+								  sd_bus_error *error) {
+	struct swaybar_sni_slot *slot = data;
+	struct swaybar_sni *sni = slot->sni;
+	int menu_id = slot->menu_id;
+	wl_list_remove(&slot->link);
+	free(slot);
+
+	int need_update;
+	sd_bus_message_read_basic(msg, 'b', &need_update);
+	if (need_update) {
+		swaybar_dbusmenu_get_layout(sni->tray->menu, menu_id);
+	}
+
+    swaybar_dbusmenu_draw(sni->tray->menu, menu_id);
+
+    sd_bus_call_method_async(sni->tray->bus, NULL, sni->service, sni->menu,
+		menu_interface, "Event", NULL, NULL, "isvu", menu_id, "opened", "y", 0, 
+		time(NULL));
+
+    sway_log(SWAY_DEBUG, "%s%s opened id %d", sni->service, sni->menu, menu_id);
+
+	return 0;
+}
+
+static void open_menu_id(struct swaybar_dbusmenu *dbusmenu, int menu_id) {
+	struct swaybar_dbusmenu_menu *menu = find_menu_id(dbusmenu->menu, menu_id);
+
+	if (!menu) {
+		return;
+	}
+
+	if (menu->surface) {
+		// Menu is already shown
+		return;
+	}
+
+	struct swaybar_sni *sni = dbusmenu->sni;
+	struct swaybar_sni_slot *slot = calloc(1, sizeof(struct swaybar_sni_slot));
+	slot->sni = sni;
+	slot->menu_id = menu_id;
+
+	int ret = sd_bus_call_method_async(sni->tray->bus, &slot->slot, sni->service,
+		sni->menu, menu_interface, "AboutToShow", about_to_show_callback, slot, "i", 
+		menu_id);
+
+	if (ret >= 0) {
+		wl_list_insert(&sni->slots, &slot->link);
+	} else {
+		sway_log(SWAY_ERROR, "%s%s failed to send AboutToShow signal: %s",
+			sni->service, sni->menu, strerror(-ret));
+		free(slot);
+	}
+}
+
+static int update_item_properties(struct swaybar_dbusmenu_menu_item *item,
+								  sd_bus_message *msg) {
+	sd_bus_message_enter_container(msg, 'a', "{sv}");
+	while (!sd_bus_message_at_end(msg, 0)) {
+		sd_bus_message_enter_container(msg, 'e', "sv");
+		char *key, *log_value;
+		sd_bus_message_read_basic(msg, 's', &key);
+		if (strcmp(key, "type") == 0) {
+			char *type;
+			sd_bus_message_read(msg, "v", "s", &type);
+			item->is_separator = strcmp(type, "separator") == 0;
+			log_value = type;
+		} else if (strcmp(key, "label") == 0) {
+			char *label;
+			sd_bus_message_read(msg, "v", "s", &label);
+			item->label = realloc(item->label, strlen(label) + 1);
+			if (!item->label) {
+				return -ENOMEM;
+			}
+			int i = 0;
+			for (char *c = label; *c; ++c) {
+				if (*c == '_' && !*++c) {
```

Comment:
Why do we skip underscores here? Without, this would have been a simple `item->label = strdup(label);`.

Reply:
From the dbusmenu protocol documentation:
```
              <tr>
                        <td>label</td>
                        <td>string</td>
                        <td>Text of the item, except that:
                        -# two consecutive underscore characters "__" are displayed as a
                        single underscore,
                        -# any remaining underscore characters are not displayed at all,
                        -# the first of those remaining underscore characters (unless it is
                        the last character in the string) indicates that the following
                        character is the access key.
                        </td>
                        <td>""</td>
                </tr>
```

Reply:
I believe a single underscore should be retained, however: only the underscores following two underscores should be elided -- that's how I read that spec.

Reply:
Add logic to display the escaped underscores (`__`) as a single underscore?  The current logic just throws away all underscores as I read it, which isn't to spec.

Reply:
I haven't written the logic it's from @ianyfan, but I believe it is correct. The `!*++c` skips the first '_'. Then if there is a second '_' it will be added to the label.

Reply:
`!*++c` means: Go to the next character. Dereference it and if it's 0 then the statement becomes true. That means doing nothing and just adding the next character to the label if we haven't reached the end of the string. In case, we reached the end of the string the loop can be terminated.

Honsetly, I'm not a fan of this code. If we have such a discussion here it should probably be simplified to make it easier to reason about.

Reply:
Ah ... sorry, too much JavaScript on my brain, not really an excuse though: I flat-out missed the inversion.
I think still though this is ignoring the spec; `____` will render as `__`, and we ignore the accelerator keys etc. -- I don't think this should hold up this effort, but there are lots of little things like this that could be improved.